File: common/src/main/maestro/pathing/movement/movements/MovementDownward.java
package maestro.pathing.movement.movements;

import com.google.common.collect.ImmutableSet;
import java.util.Set;
import maestro.api.IAgent;
import maestro.api.pathing.movement.ActionCosts;
import maestro.api.pathing.movement.MovementStatus;
import maestro.api.utils.PackedBlockPos;
import maestro.pathing.movement.CalculationContext;
import maestro.pathing.movement.Movement;
import maestro.pathing.movement.MovementHelper;
import maestro.pathing.movement.MovementState;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockState;

public class MovementDownward extends Movement {

    private int numTicks = 0;

    public MovementDownward(IAgent maestro, PackedBlockPos start, PackedBlockPos end) {
        super(maestro, start, end, new PackedBlockPos[] {end});
    }

    @Override
    public void reset() {
        super.reset();
        numTicks = 0;
    }

    @Override
    public double calculateCost(CalculationContext context) {
        return cost(context, src.getX(), src.getY(), src.getZ());
    }

    @Override
    protected Set<PackedBlockPos> calculateValidPositions() {
        return ImmutableSet.of(src, dest);
    }

    public static double cost(CalculationContext context, int x, int y, int z) {
        if (!context.allowDownward) {
            return ActionCosts.COST_INF;
        }
        if (!MovementHelper.canWalkOn(context, x, y - 2, z)) {
            return ActionCosts.COST_INF;
        }
        BlockState down = context.get(x, y - 1, z);
        Block downBlock = down.getBlock();
        if (downBlock == Blocks.LADDER || downBlock == Blocks.VINE) {
            return ActionCosts.LADDER_DOWN_ONE_COST;
        } else {
            // we're standing on it, while it might be block falling, it'll be air by the time we
            // get here in the movement
            return ActionCosts.FALL_N_BLOCKS_COST[1]
                    + MovementHelper.getMiningDurationTicks(context, x, y - 1, z, down, false);
        }
    }

    @Override
    public MovementState updateState(MovementState state) {
        super.updateState(state);
        if (state.getStatus() != MovementStatus.RUNNING) {
            return state;
        }

        if (ctx.playerFeet().toBlockPos().equals(dest.toBlockPos())) {
            return state.setStatus(MovementStatus.SUCCESS);
        } else if (!playerInValidPosition()) {
            return state.setStatus(MovementStatus.UNREACHABLE);
        }
        double diffX = ctx.player().position().x - (dest.getX() + 0.5);
        double diffZ = ctx.player().position().z - (dest.getZ() + 0.5);
        double ab = Math.sqrt(diffX * diffX + diffZ * diffZ);

        if (numTicks++ < 10 && ab < 0.2) {
            return state;
        }
        MovementHelper.moveTowards(ctx, state, positionsToBreak[0].toBlockPos(), maestro);
        return state;
    }
}

File: common/src/main/maestro/pathing/movement/movements/MovementDownward.kt
package maestro.pathing.movement.movements

import maestro.Agent
import maestro.api.pathing.movement.ActionCosts
import maestro.api.pathing.movement.MovementStatus
import maestro.api.player.PlayerContext
import maestro.api.utils.PackedBlockPos
import maestro.pathing.BlockStateInterface
import maestro.pathing.movement.CalculationContext
import maestro.pathing.movement.ClickIntent
import maestro.pathing.movement.Intent
import maestro.pathing.movement.LookIntent
import maestro.pathing.movement.Movement
import maestro.pathing.movement.MovementIntent
import maestro.pathing.movement.MovementSpeed
import maestro.pathing.movement.MovementValidation
import maestro.utils.center
import maestro.utils.centerXZ
import maestro.utils.horizontalDistanceTo
import net.minecraft.core.BlockPos
import net.minecraft.world.level.block.Blocks

/**
 * Downward movement by one block (breaking the block below if needed).
 *
 * This movement type handles:
 * - Descending one block by breaking the block below
 * - Climbing down ladders/vines
 *
 * Implementation notes:
 * - Uses pre-centering to avoid getting stuck on block edges when descending
 * - Centers horizontally before allowing gravity to pull the player down
 * - Pre-centering may not always be valid for blocks with non-typical collision boxes
 *   (e.g., open trapdoors, glass panes, iron bars, end rods)
 */
class MovementDownward(
    maestro: Agent,
    src: PackedBlockPos,
    dest: PackedBlockPos,
) : Movement(maestro, src, dest) {
    override fun toBreak(bsi: BlockStateInterface): List<BlockPos> {
        // Break the block below our feet (dest position) if it's not already passable
        if (MovementValidation.canWalkThrough(ctx, dest)) {
            return emptyList()
        }
        return listOf(dest.toBlockPos())
    }

    override fun checkCompletion() {
        if (ctx.player().blockPosition() == dest.toBlockPos()) {
            state.setStatus(MovementStatus.SUCCESS)
        } else if (ctx.player().blockPosition() != src.toBlockPos() &&
            ctx.player().blockPosition() != dest.toBlockPos()
        ) {
            state.setStatus(MovementStatus.UNREACHABLE)
        }
    }

    override fun computeIntent(ctx: PlayerContext): Intent {
        val playerPos = ctx.player().position()
        val destCenter = dest.toBlockPos().center

        // Debug: Show player position to destination line
        debug.line("player-dest", playerPos, destCenter, java.awt.Color.GREEN)
        debug.point("dest-center", destCenter, java.awt.Color.YELLOW, 0.15f)

        // Check if we need to break the block below
        val needsBreaking = !MovementValidation.canWalkThrough(ctx, dest)

        // Calculate horizontal distance from dest center
        val centerXZ = dest.toBlockPos().centerXZ
        val distXZ = playerPos.horizontalDistanceTo(centerXZ)

        // Debug: Distance and velocity metrics
        val distVertical = kotlin.math.abs(playerPos.y - dest.y)
        val velocity = ctx.player().deltaMovement
        debug.metric("dist", distVertical)
        debug.metric("distXZ", distXZ)
        debug.metric("vel", kotlin.math.abs(velocity.y))

        // Debug: Y position status
        val playerY = ctx.player().blockPosition().y
        debug.status("y", if (playerY == dest.y) "✓" else "$playerY→${dest.y}")

        // Debug: Mining and falling status
        if (needsBreaking) {
            debug.block("to-break", dest.toBlockPos(), java.awt.Color.CYAN, 0.5f)
            debug.line(
                "break-block",
                ctx.playerHead(),
                destCenter,
                java.awt.Color.RED,
            )
            debug.status("mine", "block")
        } else {
            debug.status("mine", "none")
        }

        debug.flag("break", needsBreaking)
        debug.flag("fall", velocity.y < -0.01)

        // Phase 1: Center horizontally before descending to avoid edge-sticking
        // This prevents getting stuck on block edges when falling
        if (distXZ > 0.15) {
            debug.status("phase", "center")
            return Intent(
                movement =
                    MovementIntent.Toward(
                        target = centerXZ,
                        speed = MovementSpeed.WALK,
                        startPos = src.toBlockPos().centerXZ,
                    ),
                look = LookIntent.Block(dest.toBlockPos()),
                click = if (needsBreaking) ClickIntent.LeftClick else ClickIntent.None,
            )
        }

        // Phase 2: Centered - now safe to descend
        debug.status("phase", "fall")
        return Intent(
            movement = MovementIntent.Stop, // Stay still and let gravity pull us down
            look = LookIntent.Block(dest.toBlockPos()),
            click = if (needsBreaking) ClickIntent.LeftClick else ClickIntent.None,
        )
    }

    override fun calculateCost(context: CalculationContext): Double = cost(context, src.x, src.y, src.z)

    override val validPositions: Set<PackedBlockPos>
        get() = setOf(src, dest)

    companion object {
        @JvmStatic
        fun cost(
            context: CalculationContext,
            x: Int,
            y: Int,
            z: Int,
        ): Double {
            if (!context.allowDownward) {
                return ActionCosts.COST_INF
            }
            if (!MovementValidation.canWalkOn(context, x, y - 2, z)) {
                return ActionCosts.COST_INF
            }
            val down = context[x, y - 1, z]
            val downBlock = down.block
            return if (downBlock == Blocks.LADDER || downBlock == Blocks.VINE) {
                ActionCosts.LADDER_DOWN_ONE_COST
            } else {
                // Standing on block that might be falling, but will be air by the time we get here
                ActionCosts.FALL_N_BLOCKS_COST[1] +
                    MovementValidation.getMiningDurationTicks(context, x, y - 1, z, down, false)
            }
        }
    }
}
